# Hypothesis Testing

> [!NOTE]
> To reinforce the concepts you have learned today, you will conduct a t-test on a real-world dataset. This exercise will help you practice selecting the appropriate type of t-test and interpreting the results in the context of your data. You may reuse the dataset from the previous assignment or choose a new one. Remember the [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/) is a great source for datasets. A pre-selection of candidates is provided below.
> 
> **Task Instructions**
> 1.	Select a dataset (**at least one**).
> 2.	Identify suitable samples within the dataset that allow you to test a meaningful hypothesis using a t-test.
> 3.	Decide which type of t-test is appropriate (e.g., one-sample, independent two-sample, or paired).
> 4.	Implement the t-test in Python.
> 5.	Interpret and explain the results clearly, linking them back to your hypothesis.

## Suggested Datasets

- **Palmer Penguins:** [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/dataset/690/palmer+penguins-3) | [R package](https://allisonhorst.github.io/palmerpenguins/) | [CSV export](https://gist.github.com/slopp/ce3b90b9168f2f921784de84fa445651) 
- **Wine Quality Dataset:** [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/dataset/186/wine+quality)
- **Adult Income Dataset:** [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/adult) 
- **Heart Disease Dataset:** [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/dataset/45/heart+disease) 
- **Car Evaluation Dataset:** [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/car+evaluation) 

> [!IMPORTANT]
> Please implement your solutions in a Jupyter notebook called `ttest.ipynb` and commit it to your repository.

In [11]:
# 1. Dataset selection 

from ucimlrepo import fetch_ucirepo 

# Code used from the link shown below
# Link adult dataset - https://archive.ics.uci.edu/dataset/2/adult

# fetch dataset 
adult = fetch_ucirepo(id=2) 

# data (as pandas dataframes) 
x = adult.data.features 
y = adult.data.targets 

# metadata 
print(adult.metadata) 

# variable information 
print(adult.variables) 

{'uci_id': 2, 'name': 'Adult', 'repository_url': 'https://archive.ics.uci.edu/dataset/2/adult', 'data_url': 'https://archive.ics.uci.edu/static/public/2/data.csv', 'abstract': 'Predict whether annual income of an individual exceeds $50K/yr based on census data. Also known as "Census Income" dataset. ', 'area': 'Social Science', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 48842, 'num_features': 14, 'feature_types': ['Categorical', 'Integer'], 'demographics': ['Age', 'Income', 'Education Level', 'Other', 'Race', 'Sex'], 'target_col': ['income'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 1996, 'last_updated': 'Tue Sep 24 2024', 'dataset_doi': '10.24432/C5XW20', 'creators': ['Barry Becker', 'Ronny Kohavi'], 'intro_paper': None, 'additional_info': {'summary': "Extraction was done by Barry Becker from the 1994 Census database.  A set of reasonably clean records was extracted using the fol

#### 2. "Identify suitable samples within the dataset that allow you to test a meaningful hypothesis using a t-test. "   
Choosen variables: `hours-per-week` and `income`

groups: 
- mu1 -> mean weekly working hours for individuals earning >50k
- mu2 -> mean weekly working hours for individuals earning <=50k  

hypothesis (two sided):
- H0 -> mu1 = mu2 -> There is no difference in mean weekly working hours between income groups.
- H1 -> mu1 != mu2 -> There is a difference in mean weekly working hours between income groups.
- alpha = 0.05


#### 3. "Decide which type of t-test is appropriate (e.g., one-sample, independent two-sample, or paired)."  

I use an independent two-sample t-test:  
- two independent groups (>50k and <=50k)
- Every Value represents a person that can't earn <50k and >=50k

In [15]:
# 4. Implement the t-test in Python.
import pandas as pd
from scipy.stats import ttest_ind

# Create the DataFrame
df = pd.concat([x, y], axis=1)

# Drop missing Values
df = df.dropna()

# Short visualization
print(df.groupby('income')['hours-per-week'].agg(['count', 'mean', 'std']))

# Create 2 samples
high_income_hours = df[df["income"] == ">50K"]["hours-per-week"]
low_income_hours = df[df["income"] == "<=50K"]["hours-per-week"]

# Independent 2 side t-test - no equal variance
t_value, p_value = ttest_ind(
    high_income_hours,
    low_income_hours,
    equal_var=False 
)
#t_stat, p_value
print()
print(f"t = {t_value} and p = {p_value}")

        count       mean        std
income                             
<=50K   24720  38.840210  12.318995
<=50K.  11360  39.418750  12.020953
>50K     7841  45.473026  11.012971
>50K.    3700  45.657838  10.921441

t = 45.123095093109875 and p = 0.0


#### 5. Interpret and explain the results clearly, linking them back to your hypothesis
t ~ 45.12 & p = 0

> Since the p-value is far below the chosen significance level of α = 0.05 (p < 0.001), the null hypothesis (μ₁ = μ₂) is rejected. This indicates a highly statistically significant difference in mean weekly working hours between the two income groups. The positive t-statistic suggests that individuals with higher income work more hours per week on average than those with lower income. (*Sentence corrected by ChatGPT)
